kedro/extras/datasets/svmlight/svmlight_dataset.py

```diff
@@ -0,0 +1,163 @@
+"""``SVMLightDataSet`` loads/saves data from/to a svmlight/libsvm file using an
+underlying filesystem (e.g.: local, S3, GCS). It uses sklearn functions
+``dump_svmlight_file`` to save and ``load_svmlight_file`` to load a file.
+"""
+from copy import deepcopy
+from pathlib import PurePosixPath
+from typing import Any, Dict, Optional, Tuple, Union
+
+import fsspec
+from numpy import ndarray
+from scipy.sparse.csr import csr_matrix
+from sklearn.datasets import dump_svmlight_file, load_svmlight_file
+
+from kedro.io.core import (
+    AbstractVersionedDataSet,
+    DataSetError,
+    Version,
+    get_filepath_str,
+    get_protocol_and_path,
+)
+
+_DI = Tuple[Union[ndarray, csr_matrix], ndarray]
+"""Type of the data input."""
+_DO = Tuple[csr_matrix, ndarray]
+"""Type of the data output."""
+
+
+class SVMLightDataSet(AbstractVersionedDataSet[_DI, _DO]):
+    """``SVMLightDataSet`` loads/saves data from/to a svmlight/libsvm file using an
```

Comment:
So if I understand it correctly, this dataset can be used to load and save data in both the `svmlight` and `libsvm` formats? Is there a difference in how users should use the dataset depending on the format they want to use?

Reply:
From what I've been able to find, there is no difference, both formats will be loaded and saved to `.svm` files through the [sklearn implementation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_svmlight_file.html).

Reply:
@merelcht From what I've been able to find, there is no difference, both formats will be loaded and saved to `.svm` files through the [sklearn implementation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_svmlight_file.html).

Reply:
As far as I know the `svmlight` format is an extension of `libsvm` fromat.

LIBSVM library is used for SVM classification and regression. Original `libsvm` format looks like this

```
<label> <index1>:<value1> <index2>:<value2> ...
```

In addition, SVM-Light library solves ranking tasks. Ranking problems require inputs to be grouped into query groups. So `svmlight` format also added an optional parameter for an input  - query id (qid). Inputs with the same qid belong to the same group.

```
3 qid:1 1:0.53 2:0.12
2 qid:1 1:0.13 2:0.1
7 qid:2 1:0.87 2:0.12
```

This is the only difference between them that I know.

The names for these formats are given after the libraries where they are coming from, but they are still used with other Gradient Boosting Machines libraries. For example, XGBoost extended further the format and added an [optional weight parameter to labels](https://xgboost.readthedocs.io/en/stable/tutorials/input_format.html#instance-weights)

```
1:1.0 101:1.2 102:0.03
0:0.5 1:2.1 10001:300 10002:400
```

This extension is not supported by sklearn.

Reply:
Thanks for explaining this in more detail @vilozio 🙂